### PR TITLE
Evict graph from cache upon configuration updates

### PR DESCRIPTION
**Pull request by dpitera**
When a graph configuration is updated or removed from the
ConfigurationGraphManagement, the related graph is evicted from the
cache on all JanusGraph nodes in the cluster. This is done to prevent
users from opening graphs with old/incorrect configurations across the
cluster and finding themselves in an inconsistent state. Note that this
is only done for users updating the ConfigurationManagementGraph by
going through the ConfiguredGraphFactory APIs; this is done for two
reasons: 1. to allow more fine-tuned control for users who wish to use
the lower-levls CMG APIs and 2. to avoid a cyclical dependency between
the CMG and the JanusGraphManager.

Issues: #590

Signed-off-by: David Pitera <dpitera@us.ibm.com>

-----

Thank you for contributing to JanusGraph!

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there an issue associated with this PR? Is it referenced in the commit message?
- [x] Does your PR body contain #xyz where xyz is the issue number you are trying to resolve?
- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?
- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you written and/or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the LICENSE.txt file, including the main LICENSE.txt file in the root of this repository?
- [x] If applicable, have you updated the NOTICE.txt file, including the main NOTICE.txt file found in the root of this repository?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?
- [x] If this PR is a documentation-only change, have you added a `[skip ci]`
  tag to the first line of your commit message to avoid spending CPU cycles in
  Travis CI when no code, tests, or build configuration are modified?

### Note:
Please ensure that once the PR is submitted, you check Travis CI for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementSystem.java [hunk 1]
```diff
@@ -744,6 +746,30 @@ public IndexJobFuture updateIndex(Index index, SchemaAction updateAction) {
         return future;
     }
 
+    /**
+     * Upon the open managementsystem's commit, this graph will be evicted from the cache on all JanusGraph nodes in your
+     * cluster, once there are no open transactions on this graph on each respective JanusGraph node.
```
Comment (dpitera) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementSystem.java]: and assuming each node in the cluster is configured correctly to use the JGM

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java [hunk 2]
```diff
@@ -191,12 +208,24 @@ public void run() {
                         txStillOpen = true;
                     }
                 }
-                if (!txStillOpen) {
+                final JanusGraphManager jgm = JanusGraphManager.getInstance();
+                final boolean janusGraphManagerIsInBadState = null == jgm && action.equals(EVICT);
+                if (!txStillOpen && janusGraphManagerIsInBadState) {
+                    log.error("JanusGraphManager should be instantiated on this server, but it is not. " +
```
Comment (sjudeng) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java]: Can you add a test to cover this case?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java [hunk 3]
```diff
@@ -191,12 +208,24 @@ public void run() {
                         txStillOpen = true;
                     }
                 }
-                if (!txStillOpen) {
+                final JanusGraphManager jgm = JanusGraphManager.getInstance();
+                final boolean janusGraphManagerIsInBadState = null == jgm && action.equals(EVICT);
```
Comment (sjudeng) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java]: If `jgm` is null couldn't that just mean that JanusGraph Server is not being used?

Comment (dpitera) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java]: The only way `action.equals(EVICT)` is if the user actively called `mgmt.evictGraphFromCache()` on their graph's mangementSystem, or if it was called by the ConfiguredGraphFactory [updateConfiguration](https://github.com/dpitera/janusgraph/blob/06d519e48dd64ae8df4dda5c1e94ca34d2eac250/janusgraph-core/src/main/java/org/janusgraph/core/ConfiguredGraphFactory.java#L203) or [removeConfiguration](https://github.com/dpitera/janusgraph/blob/06d519e48dd64ae8df4dda5c1e94ca34d2eac250/janusgraph-core/src/main/java/org/janusgraph/core/ConfiguredGraphFactory.java#L229). 

The issue in this scenario is indeed that a user is using the JanusGraph Server (more specifically configured to use the JanusGraphManager) for one of their JG nodes but not the other node (the one that _received_ this eviction request).

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java (hunk 2 above)
Comment (dpitera) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java]: Note that this would be hit asynchronously on a node other than the one the original action was made on, so this is a log action-- not the throwing of an error. I can add a test that ensures if the JGM is null that we don't remove it from the cache, but that is pretty obvious.

In terms of testing this functionality across multi-node deployments, I would like to avoid if we can as it would be ~non-trivial as there is no infra in place to do so, and there is also no infra in place to test the existing ManagementLogger.

Comment (sjudeng) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java]: ManagementLogger seems to be otherwise well covered in tests (see Codecov). If the test you describe hits this case I think it's worth including.

Comment (dpitera) [reply janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java]: That code-coverage analysis is pretty cool!

I can hit the case with the simpler test of opening a graph without configuring the JGM and trying to evict the graph from the cache.

**Comment by dpitera** [comment]
@sjudeng Pushed new code with your suggested updates and I added 2 tests, both testing the graph management logger's graph cache eviction logic; one when the JGM is null (mostly an obvious test, but we will see the code get covered in the codecov analysis) and one when it is not (the test proving the functionality works, i.e. the log can successfully evict graphs from the cache).

**Comment by dpitera** [comment]
> Also do you think we'll eventually want to move JanusGraphManager to the janusgraph-server module?

Sorry, I forgot to answer this question. Artificially, _maybe_ the JanusGraphManager should live in the `janusgraph-server` module. I think it's easy to sound like this is the right place to live, especially when one considers that the notion of a GraphManager is a TinkerPop ideation. Concretely, this gets complicated because things like the JanusGraphFactory, ConfiguredGraphFactory, things that live in the `janusgraph-core` module... rely on the JanusGraphManager; this would cause a cyclical dependency between the two modules. Even theoretically I am more of the notion that the `core` project is indeed the best place for the JGM to reside. What do you think? Might not be a bad thing over which to open a discussion. Either way, if the community deems it is best to move, then it would be best done in a separate PR with its own refactoring issue.

**Comment by sjudeng** [comment]
Thanks for the explanation. It looks to me like ConfiguredGraphFactory could maybe be moved over as is. I don't see any integration to other classes in core. JGM would be more painful given integration in JanusGraphFactory and (with this PR) ManagementLogger. Can we create a separate issue to investigate further, maybe specifically for moving over ConfiguredGraphFactory if you agree that would be straight-forward, as well as for identifying/moving any other relevant classes if any?

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java [hunk 4]
```diff
@@ -191,12 +211,24 @@ public void run() {
...
+                if (!txStillOpen && janusGraphManagerIsInBadState) {
```
Comment (sjudeng) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java]: Looks like this branch still isn't getting hit (see [here](https://codecov.io/gh/JanusGraph/janusgraph/src/graph-cache-evictions/janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java#L214...221)). It's not a big deal.

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java [hunk 5]
```diff
@@ -30,12 +30,16 @@
...
+import org.janusgraph.graphdb.management.JanusGraphManager;
 import org.janusgraph.graphdb.types.vertices.JanusGraphSchemaVertex;
+import static org.janusgraph.graphdb.database.management.GraphCacheEvictionAction.EVICT;
+import static org.janusgraph.graphdb.database.management.GraphCacheEvictionAction.DO_NOT_EVICT;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.time.Duration;
+import java.util.HashSet;
```
Comment (pluradj) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/ManagementLogger.java]: this isn't used

File: docs/configuredgraphfactory.adoc [hunk 6]
```diff
@@ -83,7 +83,7 @@ for which you have instantiated _and_ the references are stored inside the `Janu
 
 IMPORTANT: This is an irreversible operation that will delete all graph and index data.
 
-IMPORTANT: To ensure all graph representations are consistent across all JanusGraph nodes in your cluster, remove the graph from the `JanusGraphManager` graph reference tracker on all nodes in your cluster: `ConfiguredGraphFactory.close("graphName");`.
+IMPORTANT: To ensure all graph representations are consistent across all JanusGraph nodes in your cluster, this removes the graph from the `JanusGraphManager` graph cache on every node in the cluster, assuming each node has been properly configured to use the `JanusGraphManager`.
```
[the remark this answers is not in the document]
Comment (dpitera) [reply docs/configuredgraphfactory.adoc]: I think this suggestion makes alot of sense; I prefer to leave the call out and link to the new section about using a multinode JG cluster. I also think the information in this PR [here](https://github.com/JanusGraph/janusgraph/pull/884#issuecomment-368149520) will be relevant, so I suggest creating a _separate_ PR with this new doc section that includes the information relevant to both PRs; this will avoid merge-conflicts between these two PRs.

[the remark this answers is not in the document]
Comment (pluradj) [reply docs/configuredgraphfactory.adoc]: separate PR for docs update is fine, see https://github.com/JanusGraph/janusgraph/issues/938

File: janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/GraphCacheEvictionAction.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.janusgraph.graphdb.database.management;
+
+public enum GraphCacheEvictionAction {
+
+    EVICT,
```
Comment (amcp) [inline janusgraph-core/src/main/java/org/janusgraph/graphdb/database/management/GraphCacheEvictionAction.java]: consider documenting each of the enum values with a Javadoc.
